File: code/ts/modular_api/src/core/schema/field.ts
// ============================================================
// core/schema/field.ts
// Stage 3 decorators that store OpenAPI-compatible field metadata.
//
// Usage:
//   class HelloInput extends Input {
//     @Field.string({ description: 'Name to greet' })
//     name!: string;
//
//     @Field.optional(Field.integer({ description: 'Age' }))
//     age?: number;
//   }
//
// At runtime, `getFieldMetadata(HelloInput)` returns the list of
// FieldMeta entries in declaration order, which Input.toSchema()
// uses to build an OpenAPI 3.0.3-compatible JSON Schema.
// ============================================================

// Polyfill for runtimes that lack Symbol.metadata (Node < 22)
declare global {
  interface SymbolConstructor {
    readonly metadata: unique symbol;
  }
}
(Symbol as unknown as Record<string, unknown>).metadata ??= Symbol('Symbol.metadata');

/** Metadata stored per decorated field. */
export interface FieldMeta {
  name: string;
  type: 'string' | 'integer' | 'number' | 'boolean' | 'array' | 'object';
  description?: string;
  required: boolean;
  nullable: boolean;
  items?: { type: string };
  example?: unknown;
}

/** Options accepted by every primitive Field decorator. */
export interface FieldOptions {
  description?: string;
  example?: unknown;
}

// ── Internal storage key ─────────────────────────────────────

const FIELD_KEY = Symbol('FieldMeta');

// ── Stage 3 class field decorator factory ────────────────────

type FieldDecorator = (
  value: undefined,
  context: ClassFieldDecoratorContext,
) => void;

function makeFieldDecorator(
  type: FieldMeta['type'],
  options: FieldOptions = {},
  extra: Partial<FieldMeta> = {},
): FieldDecorator {
  return function (_value: undefined, context: ClassFieldDecoratorContext): void {
    const meta: FieldMeta = {
      name: String(context.name),
      type,
      description: options.description,
      required: true,
      nullable: false,
      ...extra,
    };
    if (options.example !== undefined) {
      meta.example = options.example;
    }

    const metadata = context.metadata as Record<symbol, FieldMeta[]>;
    if (!metadata[FIELD_KEY]) {
      metadata[FIELD_KEY] = [];
    }
    metadata[FIELD_KEY].push(meta);
  };
}

// ── Wrapper for optional fields ──────────────────────────────

/** Internal marker returned by primitive Field methods for use with Field.optional(). */
interface PendingField {
  __pending: true;
  type: FieldMeta['type'];
  options: FieldOptions;
  extra: Partial<FieldMeta>;
}

// ── Public API ───────────────────────────────────────────────

export const Field = {
  string(options: FieldOptions = {}): FieldDecorator & PendingField {
    const decorator = makeFieldDecorator('string', options);
    return Object.assign(decorator, { __pending: true as const, type: 'string' as const, options, extra: {} });
  },

  integer(options: FieldOptions = {}): FieldDecorator & PendingField {
    const decorator = makeFieldDecorator('integer', options);
    return Object.assign(decorator, { __pending: true as const, type: 'integer' as const, options, extra: {} });
  },

  number(options: FieldOptions = {}): FieldDecorator & PendingField {
    const decorator = makeFieldDecorator('number', options);
    return Object.assign(decorator, { __pending: true as const, type: 'number' as const, options, extra: {} });
  },

  boolean(options: FieldOptions = {}): FieldDecorator & PendingField {
    const decorator = makeFieldDecorator('boolean', options);
    return Object.assign(decorator, { __pending: true as const, type: 'boolean' as const, options, extra: {} });
  },

  object(options: FieldOptions = {}): FieldDecorator & PendingField {
    const decorator = makeFieldDecorator('object', options);
    return Object.assign(decorator, { __pending: true as const, type: 'object' as const, options, extra: {} });
  },

  /**
   * Marks a field as optional (nullable, not required).
   *
   * ```ts
   * @Field.optional(Field.string({ description: 'Nickname' }))
   * nickname?: string;
   * ```
   */
  optional(inner: PendingField): FieldDecorator {
    return makeFieldDecorator(inner.type, inner.options, {
      ...inner.extra,
      required: false,
      nullable: true,
    });
  },

  /**
   * Array field with typed items.
   *
   * ```ts
   * @Field.array(Field.string(), { description: 'tags' })
   * tags!: string[];
   * ```
   */
  array(itemType: PendingField, options: FieldOptions = {}): FieldDecorator & PendingField {
    const extra: Partial<FieldMeta> = { items: { type: itemType.type } };
    const decorator = makeFieldDecorator('array', options, extra);
    return Object.assign(decorator, { __pending: true as const, type: 'array' as const, options, extra });
  },
};

// ── Public accessor ──────────────────────────────────────────

/**
 * Retrieves the `@Field` metadata registered on a class.
 * Returns an empty array if the class has no decorated fields.
 */
export function getFieldMetadata(target: abstract new (...args: unknown[]) => unknown): FieldMeta[] {
  const metadata = (target as unknown as Record<symbol, Record<symbol, FieldMeta[]>>)[Symbol.metadata];
  return metadata?.[FIELD_KEY] ?? [];
}

File: code/dart/modular_api/lib/src/core/schema/field.dart
/// Schema field metadata for automatic `toSchema()` generation.
///
/// Dart does not have runtime reflection for annotations.
/// Instead of macros (experimental in Dart 3.5+), we use a declarative
/// `schemaFields` getter on Input/Output that lists field metadata.
/// The base class `Input.toSchema()` / `Output.toSchema()` builds the
/// OpenAPI 3.0.3-compatible JSON Schema automatically from this list.
///
/// ```dart
/// class HelloInput extends Input {
///   final String name;
///
///   HelloInput({required this.name});
///
///   @override
///   List<SchemaField> get schemaFields => [
///     SchemaField.string('name', description: 'Name to greet'),
///   ];
///   // toSchema() → auto-derived from schemaFields!
/// }
/// ```
library;

/// Metadata for a single schema field.
///
/// Used by [Input.toSchema] and [Output.toSchema] to build
/// OpenAPI 3.0.3-compatible JSON Schemas automatically.
class SchemaField {
  final String name;
  final String type;
  final String? description;
  final bool required;
  final bool nullable;
  final SchemaField? items;
  final dynamic example;

  const SchemaField(
    this.name,
    this.type, {
    this.description,
    this.required = true,
    this.nullable = false,
    this.items,
    this.example,
  });

  factory SchemaField.string(String name,
          {String? description, dynamic example}) =>
      SchemaField(name, 'string', description: description, example: example);

  factory SchemaField.integer(String name,
          {String? description, dynamic example}) =>
      SchemaField(name, 'integer', description: description, example: example);

  factory SchemaField.number(String name,
          {String? description, dynamic example}) =>
      SchemaField(name, 'number', description: description, example: example);

  factory SchemaField.boolean(String name,
          {String? description, dynamic example}) =>
      SchemaField(name, 'boolean', description: description, example: example);

  factory SchemaField.array(
    String name,
    SchemaField itemType, {
    String? description,
    dynamic example,
  }) =>
      SchemaField(name, 'array',
          description: description, items: itemType, example: example);

  factory SchemaField.object(String name,
          {String? description, dynamic example}) =>
      SchemaField(name, 'object', description: description, example: example);

  factory SchemaField.optional(SchemaField inner) => SchemaField(
        inner.name,
        inner.type,
        description: inner.description,
        required: false,
        nullable: true,
        items: inner.items,
        example: inner.example,
      );
}

/// Builds an OpenAPI 3.0.3 JSON Schema from a list of [SchemaField] entries.
Map<String, dynamic> buildSchema(List<SchemaField> fields) {
  final properties = <String, Map<String, dynamic>>{};
  final required = <String>[];

  final exampleValues = <String, dynamic>{};

  for (final field in fields) {
    final prop = <String, dynamic>{'type': field.type};
    if (field.description != null) {
      prop['description'] = field.description;
    }
    if (field.nullable) {
      prop['nullable'] = true;
    }
    if (field.items != null) {
      prop['items'] = {'type': field.items!.type};
    }
    if (field.example != null) {
      prop['example'] = field.example;
      exampleValues[field.name] = field.example;
    }
    properties[field.name] = prop;

    if (field.required) {
      required.add(field.name);
    }
  }

  final schema = <String, dynamic>{
    'type': 'object',
    'properties': properties,
  };
  if (required.isNotEmpty) {
    schema['required'] = required;
  }
  if (exampleValues.isNotEmpty) {
    schema['example'] = exampleValues;
  }
  return schema;
}

/// Infers an OpenAPI 3.0.3 JSON Schema from example values in a `toJson()` map.
///
/// Maps Dart runtime types to OpenAPI types:
///   - `String` → `string`
///   - `int` → `integer`
///   - `double` → `number`
///   - `bool` → `boolean`
///   - `List` → `array` (item type inferred from first element)
///   - `null` → `string` + `nullable: true` (excluded from required)
///
/// Used as fallback when [schemaFields] is not provided on an Input/Output.
Map<String, dynamic> inferSchemaFromExample(Map<String, dynamic> exampleJson) {
  final properties = <String, Map<String, dynamic>>{};
  final required = <String>[];

  for (final entry in exampleJson.entries) {
    final value = entry.value;
    if (value == null) {
      properties[entry.key] = {'type': 'string', 'nullable': true};
      continue;
    }

    final prop = <String, dynamic>{'type': _inferOpenApiType(value)};
    if (value is List) {
      prop['items'] = {
        'type': value.isEmpty ? 'string' : _inferOpenApiType(value.first),
      };
    }
    properties[entry.key] = prop;
    required.add(entry.key);
  }

  final schema = <String, dynamic>{
    'type': 'object',
    'properties': properties,
  };
  if (required.isNotEmpty) {
    schema['required'] = required;
  }
  return schema;
}

/// Maps a Dart runtime value to its OpenAPI 3.0.3 type string.
String _inferOpenApiType(dynamic value) {
  if (value is String) return 'string';
  if (value is int) return 'integer';
  if (value is double) return 'number';
  if (value is bool) return 'boolean';
  if (value is List) return 'array';
  if (value is Map) return 'object';
  return 'string';
}

/// Thrown by [validateJsonFields] when a required field is missing
/// or has the wrong JSON type.
///
/// Error messages follow the cross-SDK parity contract:
///   - `"Missing required field: {name}"`
///   - `"Field '{name}' must be of type {type}"`
class InputValidationException implements Exception {
  final String message;

  InputValidationException(this.message);

  @override
  String toString() => 'InputValidationException: $message';
}

/// Validates raw JSON against a list of [SchemaField] entries.
///
/// Throws [InputValidationException] when a required field is missing
/// or has the wrong JSON type. The handler catches this and returns 400.
///
/// Business-rule validation belongs in `UseCase.validate()` instead.
void validateJsonFields(Map<String, dynamic> json, List<SchemaField> fields) {
  for (final field in fields) {
    if (!field.required) continue;

    if (!json.containsKey(field.name) || json[field.name] == null) {
      throw InputValidationException('Missing required field: ${field.name}');
    }

    if (!_isJsonTypeValid(json[field.name], field.type)) {
      throw InputValidationException(
        "Field '${field.name}' must be of type ${field.type}",
      );
    }
  }
}

/// Checks whether a JSON value matches the expected OpenAPI type.
bool _isJsonTypeValid(dynamic value, String expectedType) {
  switch (expectedType) {
    case 'string':
      return value is String;
    case 'integer':
      return value is int;
    case 'number':
      return value is num;
    case 'boolean':
      return value is bool;
    case 'array':
      return value is List;
    case 'object':
      return value is Map;
    default:
      return true;
  }
}
